Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java

```diff
@@ -4459,4 +4495,95 @@ private static int homeBlockDistance(Town town1, Town town2) {
 
 		return (int) MathUtil.distance(town1.getHomeBlockOrNull().getCoord(), town2.getHomeBlockOrNull().getCoord());
 	}
+
+	private void parseTownForSaleCommand(Player player, String[] args) throws TownyException {
+		if (args.length == 0)
+			throw new TownyException(Translatable.of("msg_error_must_be_num"));
+
+		double forSalePrice = MathUtil.getDoubleOrThrow(args[0]);
+		Town town = getTownFromPlayerOrThrow(player);
+
+		Confirmation
+			.runOnAccept(() -> {
+				setTownForSale(town, forSalePrice, false);
+				TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_town_forsale", town.getName(), prettyMoney(forSalePrice)));
+			})
+			.setTitle(Translatable.of("msg_town_sell_confirmation", prettyMoney(forSalePrice)))
+			.sendTo(player);
+	}
+
+	private void parseTownNotForSaleCommand(Player player) throws TownyException {
+		Town town = getTownFromPlayerOrThrow(player);
+
+		if (!town.isForSale())
+			throw new TownyException(Translatable.of("msg_town_buytown_not_forsale"));
+		
+		setTownNotForSale(town, false);
+		TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_town_notforsale", town.getName()));
+	}
+
+	public static void setTownForSale(Town town, double price, boolean admin) {
+		if (town != null) {
+			town.setForSale(true);
+			town.setForSalePrice(price);
+			town.save();
+		}
+	}
+
+	public static void setTownNotForSale(Town town, boolean admin) {
+		if (town != null) {
+			town.setForSale(false);
+			town.save();
+		}
+	}
+
+	private void parseTownBuyTownCommand(CommandSender sender, String[] args) throws TownyException {
+		catchConsole(sender);
+		Player player = (Player) sender;
+
+		if (args.length == 0) {
+			throw new TownyException(Translatable.of("msg_specify_name"));
+		}
+
+		Town town = getTownOrThrow(args[0]);
+
+		if (!town.isForSale()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_not_forsale"));
+		}
+		
+		Resident resident = getResidentOrThrow(player);
+		if (resident.isMayor()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", resident.getTownOrNull().getName()));
+		}
+
+		Confirmation
+			.runOnAccept(() -> {
+				Resident currentMayor = town.getMayor();
+				

```

Comment:
This is a scenario where we shouldn't be using the `Confirmation.setCost(new ConfirmationTransaction(town::getForSalePrice, resident.getAccount(), "Town purchase cost."))` but instead use a more traditional money-check-and-pay in the Confirmation runnable to improve readability and prevent a resident paying when they cannot.

The runnable should go through three steps:
1. Test of the town is still for sale.
2. Test if the resident is already mayor.
3. Test if the resident can pay.
```java
if (!resident.getAccount().withdraw(cost)) {
    TownyMessaging.sendErrorMessage(resident, Translatable.of("YOU CANNOT AFFORD THIS"); <--- en-US.yml will have a string already in it.
    return;
}
```
Then it should begin to change the mayor over and do the rest of the work.


Afterwards, I would like you to test this out with teh following steps:

1. Resident A (mayor) sets the town for sale for $100.
5. Resident B uses the buytown command but does not confirm.
6. Resident A sets Resident B to be the mayor manually using /t set mayor.
7. Resident B hits the confirm.

Reply:
from your test

![image](https://github.com/TownyAdvanced/Towny/assets/49851457/9b7d80b5-9eee-4899-8f61-eb0151bd4d24)